File: cores/Rust/compositron/src/dbs/fitting.rs
use std::collections::HashMap;

use nalgebra::DVector;
use statrs::function::erf::erf;
use thiserror::Error;
use varpro::prelude::*;
use varpro::problem::*;
use varpro::solvers::levmar::LevMarSolver;

const TWO_OVER_SQRT_PI: f64 = 1.1283791670955126;

#[derive(Debug, Error)]
pub enum FitError {
    #[error("Data contains only NaN values")]
    AllNANValues,

    #[error("ModelBuilderError")]
    ModelBuilderError {
        #[from]
        inner: varpro::model::builder::error::ModelBuildError,
    },

    #[error("SeparableProblemBuilderError")]
    SeparableProblemBuilderError {
        #[from]
        inner: varpro::problem::SeparableProblemBuilderError,
    },

    #[error("StatisticsError")]
    StatisticsError {
        #[from]
        inner: varpro::statistics::Error<varpro::model::errors::ModelError>,
    },

    #[error("Something went wrong during least squares fit")]
    RuntimeError,
}

#[derive(Debug)]
pub struct Param {
    pub val: f64,
    pub err: f64,
}

fn gauss_basis(
    x: &DVector<f64>, x0: f64, sig: f64
) -> DVector<f64> {
    x.map(|x| {
        let u = (x - x0) / sig;
        (-0.5 * u*u).exp()
    })
}

fn gaussian_dx0(
    x: &DVector<f64>, x0: f64, sig: f64
) -> DVector<f64> {
    x.map(|x| {
        let u = (x - x0) / sig;
        (-0.5 * u*u).exp() * u / sig
    })
}

fn gaussian_dsig(
    x: &DVector<f64>, x0: f64, sig: f64
) -> DVector<f64> {
    x.map(|x| {
        let u = (x - x0) / sig;
        (-0.5 * u*u).exp() * u * u / sig
    })
}

pub fn fit_gaussian(
    x: &[f64], y: &[f64],
) -> Result<HashMap<&'static str, Param>, FitError> {
    let model = SeparableModelBuilder::<f64>::new(&["x0", "sig"])
        .function(&["x0", "sig"], gauss_basis)
        .partial_deriv("x0", gaussian_dx0)
        .partial_deriv("sig", gaussian_dsig)
        .independent_variable(DVector::from_column_slice(x))
        .initial_parameters(vec![511., 1.])
        .build()?;

    let problem = SeparableProblemBuilder::new(model)
        .observations(DVector::from_column_slice(y))
        .build()?;

    let Ok(fit_result) = LevMarSolver::default().solve(problem) else {
        return Err(FitError::RuntimeError);
    };

    let nonlin = fit_result.nonlinear_parameters();

    let Some(lin) = fit_result.linear_coefficients() else {
        return Err(FitError::RuntimeError);
    };

    let stats = varpro::statistics::FitStatistics::try_from(&fit_result)?;

    let cov = stats.covariance_matrix();

    let errs = cov.diagonal().map(|x| x.sqrt());

    let mut params = HashMap::<&'static str, Param>::new();

    params.insert("amp_1", Param { val: lin[0], err: errs[0] });
    params.insert("x0_1", Param { val: nonlin[0], err: errs[1] });
    params.insert("sig_1", Param { val: nonlin[1], err: errs[1] });

    Ok(params)
}

fn erf_basis(
    x: &DVector<f64>, x0: f64, sig: f64,
) -> DVector<f64> {
    x.map(|x| {
        let u = (x - x0) / sig;
        erf(u)
    })
}

fn erf_dx0(
    x: &DVector<f64>, x0: f64, sig: f64,
) -> DVector<f64> {
    x.map(|x| {
        let u = (x - x0) / sig;
        let e = (-u * u).exp();
        -TWO_OVER_SQRT_PI * e / sig
    })
}

fn erf_dsig(
    x: &DVector<f64>, x0: f64, sig: f64,
) -> DVector<f64> {
    x.map(|x| {
        let u = (x - x0) / sig;
        let e = (-u * u).exp();
        TWO_OVER_SQRT_PI * e * u / sig
    })
}

fn linear_basis(x: &DVector<f64>) -> DVector<f64> {
    x.clone()
}

fn const_basis(x: &DVector<f64>) -> DVector<f64> {
    DVector::from_element(x.len(), 1.)
}

/// Returns: Gauss amp, Gauss x0, Gauss sigma,
/// erf amp, linear coeff, const coeff
pub fn fit_erf_linear_1_gauss(
    x: &[f64], y: &[f64],
) -> Result<HashMap<&'static str, Param>, FitError> {
    let model = SeparableModelBuilder::<f64>::new(&["x0", "sig"])
        .function(&["x0", "sig"], gauss_basis)
        .partial_deriv("x0", gaussian_dx0)
        .partial_deriv("sig", gaussian_dsig)
        .function(&["x0", "sig"], erf_basis)
        .partial_deriv("x0", erf_dx0)
        .partial_deriv("sig", erf_dsig)
        .invariant_function(linear_basis)
        .invariant_function(const_basis)
        .independent_variable(DVector::from_column_slice(x))
        .initial_parameters(vec![511., 1.])
        .build()?;

    let weights = y
        .iter()
        .map(|&i| 1. / i.max(1.).sqrt())
        .collect::<Vec<f64>>();

    let problem = SeparableProblemBuilder::new(model)
        .observations(DVector::from_column_slice(y))
        .weights(DVector::from_vec(weights))
        .build()?;

    let Ok(fit_result) = LevMarSolver::default().solve(problem) else {
        return Err(FitError::RuntimeError);
    };

    let nonlin = fit_result.nonlinear_parameters();

    let Some(lin) = fit_result.linear_coefficients() else {
        return Err(FitError::RuntimeError);
    };

    let stats = varpro::statistics::FitStatistics::try_from(&fit_result)?;

    let cov = stats.covariance_matrix();

    let errs = cov.diagonal().map(|x| x.sqrt());

    let mut params = HashMap::<&'static str, Param>::new();

    params.insert("amp_1", Param { val: lin[0], err: errs[0] });
    params.insert("x0_1", Param { val: nonlin[0], err: errs[4] });
    params.insert("sig_1", Param { val: nonlin[1], err: errs[5] });
    params.insert("erf_amp", Param { val: lin[1], err: errs[1] });
    params.insert("lin", Param { val: lin[2], err: errs[2] });
    params.insert("const", Param { val: lin[3], err: errs[3] });

    Ok(params)
}

pub fn fit_erf_linear_2_gauss(
    x: &[f64], y: &[f64],
) -> Result<HashMap<&'static str, Param>, FitError> {
    let model = SeparableModelBuilder::<f64>::new(
            &["x0_1", "sig_1", "x0_2", "sig_2"]
        )
        .function(&["x0_1", "sig_1"], gauss_basis)
        .partial_deriv("x0_1", gaussian_dx0)
        .partial_deriv("sig_1", gaussian_dsig)
        .function(&["x0_2", "sig_2"], gauss_basis)
        .partial_deriv("x0_2", gaussian_dx0)
        .partial_deriv("sig_2", gaussian_dsig)
        .function(&["x0_1", "sig_1"], erf_basis)
        .partial_deriv("x0_1", erf_dx0)
        .partial_deriv("sig_1", erf_dsig)
        .invariant_function(linear_basis)
        .invariant_function(const_basis)
        .independent_variable(DVector::from_column_slice(x))
        .initial_parameters(vec![511., 1., 510., 1.5])
        .build()?;

    let weights = y
        .iter()
        .map(|&i| 1. / i.max(1.).sqrt())
        .collect::<Vec<f64>>();

    let problem = SeparableProblemBuilder::new(model)
        .observations(DVector::from_column_slice(y))
        .weights(DVector::from_vec(weights))
        .build()?;

    let Ok(fit_result) = LevMarSolver::default().solve(problem) else {
        return Err(FitError::RuntimeError);
    };

    let nonlin = fit_result.nonlinear_parameters();

    let Some(lin) = fit_result.linear_coefficients() else {
        return Err(FitError::RuntimeError);
    };

    let stats = varpro::statistics::FitStatistics::try_from(&fit_result)?;

    let cov = stats.covariance_matrix();

    let errs = cov.diagonal().map(|x| x.sqrt());

    let mut params = HashMap::<&'static str, Param>::new();

    params.insert("amp_1", Param { val: lin[0], err: errs[0] });
    params.insert("x0_1", Param { val: nonlin[0], err: errs[5] });
    params.insert("sig_1", Param { val: nonlin[1], err: errs[6] });
    params.insert("amp_2", Param { val: lin[1], err: errs[1] });
    params.insert("x0_2", Param { val: nonlin[2], err: errs[7] });
    params.insert("sig_2", Param { val: nonlin[3], err: errs[8] });
    params.insert("erf_amp", Param { val: lin[2], err: errs[2] });
    params.insert("lin", Param { val: lin[3], err: errs[3] });
    params.insert("const", Param { val: lin[4], err: errs[4] });

    Ok(params)
}

pub fn fit_erf_linear_3_gauss(
    x: &[f64], y: &[f64],
) -> Result<HashMap<&'static str, Param>, FitError> {
    let model = SeparableModelBuilder::<f64>::new(
            &["x0_1", "sig_1", "x0_2", "sig_2", "x0_3", "sig_3"]
        )
        .function(&["x0_1", "sig_1"], gauss_basis)
        .partial_deriv("x0_1", gaussian_dx0)
        .partial_deriv("sig_1", gaussian_dsig)
        .function(&["x0_2", "sig_2"], gauss_basis)
        .partial_deriv("x0_2", gaussian_dx0)
        .partial_deriv("sig_2", gaussian_dsig)
        .function(&["x0_3", "sig_3"], gauss_basis)
        .partial_deriv("x0_3", gaussian_dx0)
        .partial_deriv("sig_3", gaussian_dsig)
        .function(&["x0_1", "sig_1"], erf_basis)
        .partial_deriv("x0_1", erf_dx0)
        .partial_deriv("sig_1", erf_dsig)
        .invariant_function(linear_basis)
        .invariant_function(const_basis)
        .independent_variable(DVector::from_column_slice(x))
        .initial_parameters(vec![511., 1., 510., 1.5, 509., 2.0])
        .build()?;

    let weights = y
        .iter()
        .map(|&i| 1. / i.max(1.).sqrt())
        .collect::<Vec<f64>>();

    let problem = SeparableProblemBuilder::new(model)
        .observations(DVector::from_column_slice(y))
        .weights(DVector::from_vec(weights))
        .build()?;

    let Ok(fit_result) = LevMarSolver::default().solve(problem) else {
        return Err(FitError::RuntimeError);
    };

    let nonlin = fit_result.nonlinear_parameters();

    let Some(lin) = fit_result.linear_coefficients() else {
        return Err(FitError::RuntimeError);
    };

    let stats = varpro::statistics::FitStatistics::try_from(&fit_result)?;

    let cov = stats.covariance_matrix();

    let errs = cov.diagonal().map(|x| x.sqrt());

    let mut params = HashMap::<&'static str, Param>::new();

    params.insert("amp_1", Param { val: lin[0], err: errs[0] });
    params.insert("x0_1", Param { val: nonlin[0], err: errs[6] });
    params.insert("sig_1", Param { val: nonlin[1], err: errs[7] });
    params.insert("amp_2", Param { val: lin[1], err: errs[1] });
    params.insert("x0_2", Param { val: nonlin[2], err: errs[8] });
    params.insert("sig_2", Param { val: nonlin[3], err: errs[9] });
    params.insert("amp_3", Param { val: lin[2], err: errs[2] });
    params.insert("x0_3", Param { val: nonlin[2], err: errs[10] });
    params.insert("sig_3", Param { val: nonlin[3], err: errs[11] });
    params.insert("erf_amp", Param { val: lin[3], err: errs[3] });
    params.insert("lin", Param { val: lin[4], err: errs[4] });
    params.insert("const", Param { val: lin[5], err: errs[5] });

    Ok(params)
}

pub fn background(
    x: &[f64], amp: f64, x0: f64, sig: f64, lin: f64, off: f64
) -> Vec<f64> {
    x.iter().map(|x| {
        let u = (x - x0) / sig;
        amp * erf(u) + lin * x + off
    }).collect()
}

File: kernels/Python/compositron/src/compositron/dbs/fitting.py
import numpy as np
import numpy.typing as npt
from scipy.special import erf
from lmfit import Model

from ..constants import TWO_OVER_SQRT_PI
from ..core.fitting import SimpleFitParam


def gauss(
    x: npt.NDArray[np.floating],
    amp: float,
    x0: float,
    sig: float,
) -> npt.NDArray[np.floating]:
    return amp * np.exp(-0.5 * ((x - x0) / sig)**2)


def fit_gaussian(
    x: npt.NDArray[np.floating],
    y: npt.NDArray[np.floating],
    init: list[float | np.floating],
) -> dict[str, SimpleFitParam]:
    model = Model(gauss)

    def jac(params, _, weights, x):
        amp = params["amp"]
        x0 = params["x0"]
        sig = params["sig"]

        j = np.empty((3, x.size), dtype=np.float64)

        e = gauss(x, amp, x0, sig)
        u = (x - x0) / sig
        z = u / sig

        j[0, :] = e / amp
        j[1, :] = e * z
        j[2, :] = e * u * z

        j *= weights

        return j

    w = 1. / np.sqrt(np.maximum(y, 1.))

    params = model.make_params(
        amp = init[0],
        x0 = init[1],
        sig = init[2],
    )

    result = model.fit(
        y, x=x, weights=w, params=params, fit_kws={"Dfun": jac, "col_deriv": True}
    )

    params = result.params

    return {
        "amp_1": SimpleFitParam(params["amp"].value, params["amp"].stderr),
        "x0_1": SimpleFitParam(params["x0"].value, params["x0"].stderr),
        "sig_1": SimpleFitParam(params["sig"].value, params["sig"].stderr),
    }


def erf_linear_background(
    x: npt.NDArray[np.floating],
    amp: float,
    x0: float,
    sig: float,
    lin: float,
    off: float,
) -> npt.NDArray[np.floating]:
    return amp * erf((x - x0) / sig) + lin * x + off


def erf_linear_1_gauss(
    x: npt.NDArray[np.floating],
    amp_1: float,
    x0_1: float,
    sig_1: float,
    erf_amp: float,
    lin: float,
    off: float,
) -> npt.NDArray[np.floating]:
    return (
        gauss(x, amp_1, x0_1, sig_1)
        + erf_linear_background(x, erf_amp, x0_1, sig_1, lin, off)
    )


def fit_erf_linear_1_gauss(
    x: npt.NDArray[np.floating],
    y: npt.NDArray[np.floating],
    init: list[float | np.floating],
) -> dict[str, SimpleFitParam]:
    model = Model(erf_linear_1_gauss)

    def jac(params, _, weights, x):
        amp_1 = params["amp_1"]
        x0_1 = params["x0_1"]
        sig_1 = params["sig_1"]

        j = np.empty((6, x.size), dtype=np.float64)

        e = gauss(x, amp_1, x0_1, sig_1)
        u = (x - x0_1) / sig_1
        z = u / sig_1

        j[0, :] = e / amp_1
        j[1, :] = e * z - TWO_OVER_SQRT_PI * np.exp(-u**2) / sig_1
        j[2, :] = e * u * z + TWO_OVER_SQRT_PI * np.exp(-u**2) * u / sig_1
        j[3, :] = erf(u)
        j[4, :] = 1.
        j[5, :] = 0.

        j *= weights

        return j

    w = 1. / np.sqrt(np.maximum(y, 1.))

    params = model.make_params(
        amp_1 = init[0],
        x0_1 = init[1],
        sig_1 = init[2],
        erf_amp = init[3],
        lin = init[4],
        off = init[5],
    )

    result = model.fit(
        y, x=x, weights=w, params=params, fit_kws={"Dfun": jac, "col_deriv": True}
    )

    params = result.params

    return {
        "amp_1": SimpleFitParam(params["amp_1"].value, params["amp_1"].stderr),
        "x0_1": SimpleFitParam(params["x0_1"].value, params["x0_1"].stderr),
        "sig_1": SimpleFitParam(params["sig_1"].value, params["sig_1"].stderr),
        "erf_amp": SimpleFitParam(params["erf_amp"].value, params["erf_amp"].stderr),
        "lin": SimpleFitParam(params["lin"].value, params["lin"].stderr),
        "const": SimpleFitParam(params["off"].value, params["off"].stderr),
    }


def erf_linear_2_gauss(
    x: npt.NDArray[np.floating],
    amp_1: float,
    x0_1: float,
    sig_1: float,
    amp_2: float,
    x0_2: float,
    sig_2: float,
    erf_amp: float,
    lin: float,
    off: float,
) -> npt.NDArray[np.floating]:
    return (
        gauss(x, amp_1, x0_1, sig_1)
        + gauss(x, amp_2, x0_2, sig_2)
        + erf_linear_background(x, erf_amp, x0_1, sig_1, lin, off)
    )


def fit_erf_linear_2_gauss(
    x: npt.NDArray[np.floating],
    y: npt.NDArray[np.floating],
    init: list[float | np.floating],
) -> dict[str, SimpleFitParam]:
    model = Model(erf_linear_1_gauss)

    def jac(params, _, weights, x):
        amp_1 = params["amp_1"]
        x0_1 = params["x0_1"]
        sig_1 = params["sig_1"]
        amp_2 = params["amp_2"]
        x0_2 = params["x0_2"]
        sig_2 = params["sig_2"]

        j = np.empty((9, x.size), dtype=np.float64)

        e_1 = gauss(x, amp_1, x0_1, sig_1)
        e_2 = gauss(x, amp_2, x0_2, sig_2)
        u_1 = (x - x0_1) / sig_1
        u_2 = (x - x0_2) / sig_2
        z_1 = u_1 / sig_1
        z_2 = u_2 / sig_2

        j[0, :] = e_1 / amp_1
        j[1, :] = e_1 * z_1 - TWO_OVER_SQRT_PI * np.exp(-u_1**2) / sig_1
        j[2, :] = e_1 * u_1 * z_1 + TWO_OVER_SQRT_PI * np.exp(-u_1**2) * u_1 / sig_1
        j[3, :] = e_2 / amp_2
        j[4, :] = e_2 * z_2
        j[5, :] = e_2 * u_2 * z_2
        j[6, :] = erf(u_1)
        j[7, :] = 1.
        j[8, :] = 0.

        j *= weights

        return j

    w = 1. / np.sqrt(np.maximum(y, 1.))

    params = model.make_params(
        amp_1 = init[0],
        x0_1 = init[1],
        sig_1 = init[2],
        amp_2 = init[3],
        x0_2 = init[4],
        sig_2 = init[5],
        erf_amp = init[6],
        lin = init[7],
        off = init[8],
    )

    result = model.fit(
        y, x=x, weights=w, params=params, fit_kws={"Dfun": jac, "col_deriv": True}
    )

    params = result.params

    return {
        "amp_1": SimpleFitParam(params["amp_1"].value, params["amp_1"].stderr),
        "x0_1": SimpleFitParam(params["x0_1"].value, params["x0_1"].stderr),
        "sig_1": SimpleFitParam(params["sig_1"].value, params["sig_1"].stderr),
        "amp_2": SimpleFitParam(params["amp_2"].value, params["amp_2"].stderr),
        "x0_2": SimpleFitParam(params["x0_2"].value, params["x0_2"].stderr),
        "sig_2": SimpleFitParam(params["sig_2"].value, params["sig_2"].stderr),
        "erf_amp": SimpleFitParam(params["erf_amp"].value, params["erf_amp"].stderr),
        "lin": SimpleFitParam(params["lin"].value, params["lin"].stderr),
        "const": SimpleFitParam(params["off"].value, params["off"].stderr),
    }


def erf_linear_3_gauss(
    x: npt.NDArray[np.floating],
    amp_1: float,
    x0_1: float,
    sig_1: float,
    amp_2: float,
    x0_2: float,
    sig_2: float,
    amp_3: float,
    x0_3: float,
    sig_3: float,
    erf_amp: float,
    lin: float,
    off: float,
) -> npt.NDArray[np.floating]:
    return (
        gauss(x, amp_1, x0_1, sig_1)
        + gauss(x, amp_2, x0_2, sig_2)
        + gauss(x, amp_3, x0_3, sig_3)
        + erf_linear_background(x, erf_amp, x0_1, sig_1, lin, off)
    )


def fit_erf_linear_3_gauss(
    x: npt.NDArray[np.floating],
    y: npt.NDArray[np.floating],
    init: list[float | np.floating],
) -> dict[str, SimpleFitParam]:
    model = Model(erf_linear_1_gauss)

    def jac(params, _, weights, x):
        amp_1 = params["amp_1"]
        x0_1 = params["x0_1"]
        sig_1 = params["sig_1"]
        amp_2 = params["amp_2"]
        x0_2 = params["x0_2"]
        sig_2 = params["sig_2"]
        amp_3 = params["amp_3"]
        x0_3 = params["x0_3"]
        sig_3 = params["sig_3"]

        j = np.empty((12, x.size), dtype=np.float64)

        e_1 = gauss(x, amp_1, x0_1, sig_1)
        e_2 = gauss(x, amp_2, x0_2, sig_2)
        e_3 = gauss(x, amp_3, x0_3, sig_3)
        u_1 = (x - x0_1) / sig_1
        u_2 = (x - x0_2) / sig_2
        u_3 = (x - x0_3) / sig_3
        z_1 = u_1 / sig_1
        z_2 = u_2 / sig_2
        z_3 = u_3 / sig_3

        j[0, :] = e_1 / amp_1
        j[1, :] = e_1 * z_1 - TWO_OVER_SQRT_PI * np.exp(-u_1**2) / sig_1
        j[2, :] = e_1 * u_1 * z_1 + TWO_OVER_SQRT_PI * np.exp(-u_1**2) * u_1 / sig_1
        j[3, :] = e_2 / amp_2
        j[4, :] = e_2 * z_2
        j[5, :] = e_1 * u_2 * z_2
        j[6, :] = e_3 / amp_3
        j[7, :] = e_2 * z_3
        j[8, :] = e_3 * u_3 * z_3
        j[9, :] = erf(u_1)
        j[10, :] = 1.
        j[11, :] = 0.

        j *= weights

        return j

    w = 1. / np.sqrt(np.maximum(y, 1.))

    params = model.make_params(
        amp_1 = init[0],
        x0_1 = init[1],
        sig_1 = init[2],
        amp_2 = init[3],
        x0_2 = init[4],
        sig_2 = init[5],
        amp_3 = init[6],
        x0_3 = init[7],
        sig_3 = init[8],
        erf_amp = init[9],
        lin = init[10],
        off = init[11],
    )

    result = model.fit(
        y, x=x, weights=w, params=params, fit_kws={"Dfun": jac, "col_deriv": True}
    )

    params = result.params

    return {
        "amp_1": SimpleFitParam(params["amp_1"].value, params["amp_1"].stderr),
        "x0_1": SimpleFitParam(params["x0_1"].value, params["x0_1"].stderr),
        "sig_1": SimpleFitParam(params["sig_1"].value, params["sig_1"].stderr),
        "amp_2": SimpleFitParam(params["amp_2"].value, params["amp_2"].stderr),
        "x0_2": SimpleFitParam(params["x0_2"].value, params["x0_2"].stderr),
        "sig_2": SimpleFitParam(params["sig_2"].value, params["sig_2"].stderr),
        "erf_amp": SimpleFitParam(params["erf_amp"].value, params["erf_amp"].stderr),
        "lin": SimpleFitParam(params["lin"].value, params["lin"].stderr),
        "const": SimpleFitParam(params["off"].value, params["off"].stderr),
    }
